pkg/sql/schemachanger/sctest/cumulative.go

```diff
@@ -52,34 +470,57 @@ import (
 func cumulativeTest(
 	t *testing.T,
 	relPath string,
-	tf func(t *testing.T, path string, rewrite bool, setup, stmts []parser.Statement),
+	tf func(t *testing.T, path string, rewrite bool, setup, stmts []parser.Statement, stageExecMap *stageExecStmtMap),
 ) {
 	skip.UnderStress(t)
 	skip.UnderRace(t)
 	path := testutils.RewritableDataPath(t, relPath)
 	var setup []parser.Statement
-	datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
-		stmts, err := parser.Parse(d.Input)
-		require.NoError(t, err)
-		require.NotEmpty(t, stmts)
+	stageExecMap := makeStageExecStmtMap()
+	rewrite := false
+	var testStmts parser.Statements
+	var lines []string
 
+	// First pass collect information.
+	datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
 		switch d.Cmd {
 		case "setup":
-			// no-op
+			stmts, err := parser.Parse(d.Input)
+			setup = append(setup, stmts...)
+			require.NoError(t, err)
+			require.NotEmpty(t, stmts)
+		// no-op
+		case "stage-exec":
+			stageExecMap.ParseStageExec(t, d)
+		case "stage-query":
+			stageExecMap.ParseStageQuery(t, d)
 		case "test":
-			var lines []string
+			stmts, err := parser.Parse(d.Input)
+			require.NoError(t, err)
+			require.NotEmpty(t, stmts)
+			testStmts = stmts
 			for _, stmt := range stmts {
 				lines = append(lines, stmt.SQL)
 			}
-			t.Run(strings.Join(lines, "; "), func(t *testing.T) {
-				tf(t, path, d.Rewrite, setup, stmts)
-			})
-		default:
-			return fmt.Sprintf("unknown command: %s", d.Cmd)
+			rewrite = d.Rewrite
 		}
-		setup = append(setup, stmts...)
 		return d.Expected
 	})
+	t.Run(strings.Join(lines, "; "), func(t *testing.T) {
+		tf(t, path, rewrite, setup, testStmts, stageExecMap)
+	})
+	// Run through and recover the observed output for statements.
+	if rewrite {
+		datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
+			switch d.Cmd {
+			case "stage-exec":
+				fallthrough
+			case "stage-query":
+				return stageExecMap.GetExpectedOutputForPos(d.Pos)
+			}
+			return d.Expected
+		})
+	}
```

Comment:
in this cumulative variant, which we use for backup, pause, rollback testing, what should `--rewrite` do for these two newly added `stage-exec` and `stage-query` commands? Here the code seems to suggest we rewrite them with `expected output`, but then doesn't that mean nothing will change (i.e. you rewrite the output with what's already there)?

Reply:
I made it clear the expected output is only returned for all other statements. Otherwise, the map based on line position will contain the output.